synapse/storage/state.py

```diff
@@ -585,7 +649,7 @@ def _get_state_for_groups(self, groups, types=None):
                 types_to_fetch = types
```

Comment:
it looks like this is going to do the wrong thing in the presence of an unwildcarded `types` but a non-empty `filtered_types`: in particular it isn't going to fetch the event types that aren't listed in `filtered_types`.

I'd say that, if filtered_types is not None, `types_to_fetch` should be set to None.

Reply:
hopefully covered by my gist in the comment below, but if you can take a look at the gist patch before I put yet more dead ends into this PR it'd be appreciated.